packages/enhanced/src/lib/sharing/ConsumeSharedPlugin.ts

```diff
@@ -296,33 +294,69 @@ class ConsumeSharedPlugin {
 
         normalModuleFactory.hooks.factorize.tapPromise(
           PLUGIN_NAME,
-          ({ context, request, dependencies }) =>
+          async (resolveData: ResolveData): Promise<Module | undefined> => {
+            const { context, request, dependencies, contextInfo } = resolveData;
             // wait for resolving to be complete
             //@ts-ignore
-            promise.then(() => {
+            return promise.then(() => {
               if (
                 dependencies[0] instanceof ConsumeSharedFallbackDependency ||
                 dependencies[0] instanceof ProvideForSharedDependency
               ) {
                 return;
               }
+
+              // First try to match with layer-specific request
+              if (contextInfo.issuerLayer) {
+                // Try to find a layer-specific match
+                for (const [key, options] of unresolvedConsumes) {
+                  if (
+                    options.issuerLayer === contextInfo.issuerLayer &&
+                    (key === request ||
+                      (options.import && options.import === request))
```

Comment:
Why would we check `options.import` here?

Reply:
Because the key in share is like react-other etc. Since I need unique keys. So if I want to share react and react with layer, I need it to match the request and use import since the object key doesn't exist as request

Reply:
@sokra for this example:
See how `key` will not match `request` so user would have to specify `import` 
```js
    new ConsumeSharedPlugin({
      consumes: {
        react: {
          singleton: true,
        },
        'explicit-layer-react': {
          import: 'react/index2',
          shareKey: 'react',
          singleton: true,
          issuerLayer: 'differing-layer',
          layer: 'explicit-layer',
        },
        'differing-layer-react': {
          import: 'react',
          shareKey: 'react',
          singleton: true,
          layer: 'required-layer',
        },
        'layered-react': {
          import: 'react',
          shareKey: 'react',
          singleton: true,
          issuerLayer: 'other-layer',
        },
      },
    }),
```